Feature randic index

jeason15 opened the pull request:
<!--
Please run black to format your code.
See https://networkx.org/documentation/latest/developer/contribute.html for details.
-->

dschult reviewed (commented):
Just some quick comments:
- the module name and the function name should be different. Perhaps the module should have a more general name in case other measures of branching arrive.  Perhaps `branching.py`. Or maybe this is part of the literature that measures how clustered the network is, in which case it could go in the already existing `cluster.py`.
- add a newline at the end of the file.
- the code could be made more compact -- easier to read and faster by using `G.degree` directly instead of through `deg_dict`. Also, sum over an iterator instead of list and split the edge at the loop:
```python
index = sum(1/sqrt(G.degree[u] * G.degree[v]) for u, v in G.edges())
```
(I added G.edges() instead of G.edges so splitting into two nodes will work for MultiGraphs)

jeason15 commented:
> Just some quick comments:
> 
> - the module name and the function name should be different. Perhaps the module should have a more general name in case other measures of branching arrive.  Perhaps `branching.py`. Or maybe this is part of the literature that measures how clustered the network is, in which case it could go in the already existing `cluster.py`.
> 
> - add a newline at the end of the file.
> 
> - the code could be made more compact -- easier to read and faster by using `G.degree` directly instead of through `deg_dict`. Also, sum over an iterator instead of list and split the edge at the loop:
> 
> ```python
> 
> index = sum(1/sqrt(G.degree[u] * G.degree[v]) for u, v in G.edges())
> 
> ```
> 
> (I added G.edges() instead of G.edges so splitting into two nodes will work for MultiGraphs)

All comments addressed. Reformatted and renamed module. Awaiting re-review and approval.

rossbar commented on networkx/algorithms/connectivity/tests/test_molecular_connectivity.py:
```diff
@@ -0,0 +1,56 @@
+"""
+
+import networkx as nx
+import unittest
```
Networkx uses pytest instead of unittest.

It doesn't have to be a blocker as we can rewrite the tests, but just something that should be reworked at some point.

rossbar commented on networkx/algorithms/connectivity/molecular_connectivity.py:
```diff
@@ -0,0 +1,53 @@
+
+    References
+    --------
+        [1] Randic, M. (1975). Characterization of molecular branching.
+        Journal of the American Chemical Society, 97(23), 6609-6615.
```
Out of curiosity - is this publicly available without a paywall? I've never heard of the Randic index and it'd be good to be able to read the article where it was defined to verify that things are correct here and that the API is sensible.

dschult replied:
Let's use the DOI.  https://pubs.acs.org/doi/10.1021/ja00856a001
It seems to have no paywall at least for articles that old.

rossbar replied:
That's what I was looking at too... the first page is freely accessible but the article in its entirety is behind a paywall (AFAICT)

dschult replied:
Yes -- my mistake.  I think it was giving me access via my IP number on campus or something.
You are correct that it is behind a paywall.  Maybe there is another citation we can add to this one that is more recent and explains how it works.

dschult replied:
Would this article fit the bill?
https://www.tandfonline.com/doi/pdf/10.1016/j.akcej.2017.09.006
